src/message_bus.zig

```diff
@@ -18,38 +22,152 @@ pub const Message = struct {
+            .replicas = replicas,
+            .connections = connections,
         };
-        return self;
+
+        self.maybe_accept();
     }
 
-    pub fn deinit(self: *MessageBus) void {}
+    fn init_tcp(address: std.net.Address) !os.socket_t {
+        const fd = try os.socket(address.any.family, os.SOCK_STREAM | os.SOCK_CLOEXEC, 0);
+        errdefer os.close(fd);
+        // TODO: configure RCVBUF, SNDBUF, KEEPALIVE, TIMEOUT, NODELAY
+        try os.setsockopt(server, os.SOL_SOCKET, os.SO_REUSEADDR, &mem.toBytes(@as(c_int, 1)));
+        // TODO: port hopping
+        try os.bind(server, &address.any, address.getOsSockLen());
+        try os.listen(server, tcp_backlog);
+    }
+
+    /// Must be called whenever a Connection in the connections slice is closed.
+    fn maybe_accept(self: *MessageBus) void {
+        if (self.connections_used == self.connections.len) {
+            if (self.replica_connections == self.replicas.len - 1) {
+                // If there are no unusued connections and all replicas are
+                // connected, do nothing.
+                return;
+            } else {
+                assert(self.replica_connections < self.replicas.len - 1);
+                // There are no unused connections but not all replicas are connected.
+                // Disconnect a client to make sure there is space for a replica to connect.
+                log.info("all connections in use but not all replicas connected, disconnecting a client", .{});
+                for (self.connections) |connection| {
+                    assert(connection.peer != .none);
+                    if (connection.peer == .client) {
+                        connection.close();
+                        return;
+                    }
+                }
+                // This code should never be reached in normal circumstances.
+                // In the edge case that all client connections are currently
+                // waiting on the first message header to determine whether the peer
+                // is a client or a replica, disconnected one of these
+                // waiting connections to ensure we make progress.
+                log.warn("failed to disconnect a client as all peers were replicas or unknown, disconnecting an unknown peer.");
+                for (self.connections) |connection| {
+                    assert(connection.peer != .none);
+                    if (connection.peer == .unknown) {
+                        connection.close();
+                        return;
+                    }
+                }
+                // If this is reached, connections.len is too low for the
+                // current number of replicas or there is some other bug in
+                // the system.
+                unreachable;
+            }
+        }
+        assert(self.connections_used < self.connections.len);
+        self.io.accept(*MessageBus, self, accept, &server_completion, server, os.SOCK_CLOEXEC);
```

Comment:
Should we not pass `server_fd` here since I think `server` here refers to `*Replica`?